node/pkg/reporter/reporter.go

```diff
@@ -2,135 +2,298 @@
+			log.Error().Str("Player", "Reporter").Err(err).Msg("GetLatestGlobalAggregates")
+			return err
+		}
+
+		validAggregates := r.filterInvalidAggregates(aggregates)
+		if len(validAggregates) == 0 {
+			log.Error().Str("Player", "Reporter").Msg("no valid aggregates to report")
+			return nil
+		}
+
+		err = r.report(ctx, validAggregates)
+		if err != nil {
+			log.Error().Str("Player", "Reporter").Err(err).Msg("Report")
+			return err
+		}
+
+		for _, agg := range validAggregates {
+			pair := r.SubmissionPairs[agg.Name]
+			pair.LastSubmission = agg.Round
+			r.SubmissionPairs[agg.Name] = pair
+		}
+		log.Debug().Str("Player", "Reporter").Dur("duration", time.Since(start)).Msg("reporting done")
+		return nil
+	}
+
+	err := r.retry(job)
 	if err != nil {
-		return err
+		r.resignLeader()
+		log.Error().Str("Player", "Reporter").Err(err).Msg("failed to report")
+		return errors.New("failed to report")
+	}
+
+	return nil
+}
+
+func (r *Reporter) resignLeader() {
+	r.Raft.StopHeartbeatTicker()
+	r.Raft.UpdateRole(raft.Follower)
+}
+
+func (r *Reporter) handleCustomMessage(msg raft.Message) error {
+	// TODO: implement message handling related to validation
+	return errors.New("unknown message type")
+}
+
+func (r *Reporter) getLatestGlobalAggregates(ctx context.Context) ([]GlobalAggregate, error) {
+	result, err := r.getLatestGlobalAggregatesRdb(ctx)
+	if err != nil {
+		log.Error().Str("Player", "Reporter").Err(err).Msg("getLatestGlobalAggregatesRdb failed, trying to get from pgsql")
+		return r.getLatestGlobalAggregatesPgsql(ctx)
+	}
+	return result, nil
+}
+
+func (r *Reporter) getLatestGlobalAggregatesPgsql(ctx context.Context) ([]GlobalAggregate, error) {
+	names := make([]string, 0, len(r.SubmissionPairs))
+	for name := range r.SubmissionPairs {
+		names = append(names, name)
+	}
+
+	q := makeGetLatestGlobalAggregatesQuery(names)
+	return db.QueryRows[GlobalAggregate](ctx, q, nil)
+}
+
+func (r *Reporter) getLatestGlobalAggregatesRdb(ctx context.Context) ([]GlobalAggregate, error) {
+	keys := make([]string, 0, len(r.SubmissionPairs))
+
+	for name := range r.SubmissionPairs {
+		keys = append(keys, "globalAggregate:"+name)
 	}
 
-	a.subscribe(ctx)
+	result, err := db.MGet(ctx, keys)
+	if err != nil {
+		log.Error().Str("Player", "Reporter").Err(err).Msg("failed to get latest global aggregates")
+		return nil, err
+	}
+
+	aggregates := make([]GlobalAggregate, 0, len(result))
+	for i, agg := range result {
+		if agg == nil {
+			log.Error().Str("Player", "Reporter").Str("key", keys[i]).Msg("missing aggregate")
+			continue
+		}
+		var aggregate GlobalAggregate
+		err = json.Unmarshal([]byte(agg.(string)), &aggregate)
+		if err != nil {
+			log.Error().Str("Player", "Reporter").Err(err).Str("key", keys[i]).Msg("failed to unmarshal aggregate")
+			continue
+		}
+		aggregate.Name = strings.TrimPrefix(keys[i], "globalAggregate:")
+		aggregates = append(aggregates, aggregate)
+	}
 
-	return a.startReporter(ctx)
+	return aggregates, nil
```

Comment:
since getting from getting from pgsql is a backup logic it is intended behavior, I don't think it is necessary to try out both logics and combine errors. but thank you for the suggestion